### ISSUE TITLE
[HTTP/3] Support Single-file on Windows

### ISSUE THREAD

**Issue by karelz** [dotnet/runtime#73290]
HTTP/3 should be supported also in Single-file scenario on Windows.

Linux is simple as msquic is "part of OS" - external package that is prereq for HTTP/3 to work.
On Windows, we build and include msquic.dll as part of .NET Runtime.

@vitek-karas can point us to what needs to happen to make it work.
cc @jkotas

**Comment by karelz** [comment]
Triage: We discussed with @vitek-karas that such work should be part of 8.0 -- it is too late for 7.0 and the usage of HTTP/3 is not that much widespread yet.
Moreover, there is workaround (which we can document on demand) to copy msquic.dll next to Single-file (while not perfect, it is at least functional).

**Comment by vitek-karas** [comment]
@VSadov should know the technical details, but high level:
* `msquic` needs to be statically linked into the `singlefilehost.exe` during our build
* The native entry points which `msquic` exposes and are used by managed code need to be "marked" as internal, so that interop redirects them into the exe (and won't try to load `msquic.dll`
* The managed code which uses `msquic` might need changes - the above will work without changes if we use simple PInvokes (`DllImport`) but if there's other usage we might need to make some tweaks

There's obviously a size aspect to all of this - as this will make the single-file host cca 500KB bigger - for everybody.

**Comment by teo-tsirpanis** [comment]
I wonder if we could create bespoke single-file hosts that link to custom static libraries at publish-time. The apps that don't use `HttpClient` won't have to bear the burden of half a megabyte of code they won't use, and these that do use it but don't need HTTP/3 would be able to remove support for it with a feature switch, and other libraries could use it to bundle their own native dependencies. I think something similar is done by Blazor WebAssembly to reduce size.

Of course it would be an optional feature since it would need additional dependencies such as a native linker, and the default would still be a single-file host that links everything we might need.

**Comment by vitek-karas** [comment]
So far we've been trying to avoid dependency on platform linker for these scenarios. I think we might be able to do this for the NativeAOT version of this problem, since that one already has a dependency on platform linker.

**Comment by karelz** [comment]
@jkotas we build Windows in our fork https://github.com/dotnet/msquic -- that's how Windows DLLs are built for .NET Runtime product.
I assume that building and publishing .lib should not be a big problem in the same repo. cc @wfurt

**Comment by jkotas** [comment]
Yes, it is certainly doable. If the .lib is built in one repo and consumed in second repo, there are details you need to be careful about - like making sure that the compiler versions and compiler options are the same or at least compatible between the repos. It may be challenging to manage at times when the compilers are getting upgraded.

**Comment by nibanks** [comment]
Another, completely orthogonal path could be that we align the Windows model with Linux, and we ship MsQuic via a package manager ([winget](https://github.com/microsoft/winget-cli#readme)), and .NET doesn't package msquic at all. .NET already has the code to opportunistically load msquic when available.

The obvious downside with this approach is that you will always require an additional step, even in the single-file scenarios, to install the msquic dependency (if not already). So I'm not quite sure if you'd consider this solving the single-file problem or not. Generally, this seems to be deemed an acceptable model on Linux, but is not commonplace on Windows, and I suspect would receive push back. Just a thought though....

**Comment by ManickaP** [comment]
Part of Windows platform test is disabled against this issue:
https://github.com/dotnet/runtime/blob/47071da67320985a10f4b70f50f894ab411f4994/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicPlatformDetectionTests.cs#L23-L38

See #81581

**Comment by karelz** [comment]
Pushing to 9.0 (Future at this moment) per offline discussion.

**Comment by teo-tsirpanis** [comment]
Moving to 9.0.0 milestone now that it exists, per above comment.

**Comment by ladeak** [comment]
I have a .NET 8 AOT application that uses HTTP/3 with msquic on Windows x64. When I run the application 'standalone' from a console it can use H3. However, when I use the same AOT compiled code from a node application (a VSCode plugin) with (https://github.com/microsoft/node-api-dotnet), I get the following exception:

```
Request Error System.Net.Http.HttpRequestException: Requesting HTTP version 3.0 with version policy RequestVersionExact while unable to establish HTTP/3 connection.
   at System.Net.Http.HttpConnectionPool.ThrowGetVersionException(HttpRequestMessage, Int32, Exception) + 0xec
   at System.Net.Http.HttpConnectionPool.<SendWithVersionDetectionAndRetryAsync>d__89.MoveNext() + 0x572
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x20
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0xb2
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x4b
   at System.Net.Http.RedirectHandler.<SendAsync>d__4.MoveNext() + 0x18b
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x20
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0xb2
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x4b
   at System.Net.Http.HttpClient.<<SendAsync>g__Core|83_0>d.MoveNext() + 0x3b1
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x20
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0xb2
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x4b
   at CHttp.HttpMessageSender.<SendRequestAsync>d__7.MoveNext() + 0x189
```

`QuicConnection.IsSupported` returns false, that gives me the idea that loading the native library does not work.

Hence, I looked further on how the library is being [loaded](https://source.dot.net/#System.Net.Quic/System/Net/Quic/Internal/MsQuicApi.cs,271a13c14ffb04c7). I tried to call the same `NativeLibrary.TryLoad` from my application explicitly by invoking:

```csharp
loaded = NativeLibrary.TryLoad(Interop.Libraries.MsQuic, typeof( typeof(System.Net.Quic.QuicConnection).Assembly, DllImportSearchPath.AssemblyDirectory, out msQuicHandle);
```

I tried a few combinations for the library name and `DllImportSearchPath` but all returns false. However, when I try load it with a full path, it returns true: 

```csharp
NativeLibrary.TryLoad("C:\\Windows\\System32\\drivers\\msquic.sys", typeof(System.Net.Quic.QuicConnection).Assembly, DllImportSearchPath.AssemblyDirectory, out msQuicHandle);
```

Is there a way to override the default of loading the library?

**Comment by jkotas** [comment]
You need to manually copy msquic.dll next to your single-file or native-aot compiled binary as a workaround for this issue.

**Comment by ladeak** [comment]
Thank you for the suggestion

**Comment by MichalStrehovsky** [comment]
@ladeak are you on .NET 7? Native AOT had a bug that it didn't do anything for `DllImportSearchPath.AssemblyDirectory` that is used to load the msquic library. This was fixed in .NET 8 (https://github.com/dotnet/runtime/pull/90120).

Is your VS code extension an EXE or a native DLL? If it's a DLL, additional considerations apply (#90131).

**Comment by ladeak** [comment]
I am using .NET 8 RC2. The plugin is an exe that is packaged as a .node file.

**Comment by MichalStrehovsky** [comment]
The DLLs loaded with `DllImportSearchPath.AssemblyDirectory` (which is the case of MSQuic) will be loaded from wherever `AppContext.BaseDirectory` points to. This should default to the same path as `Environment.ProcessPath`. What is your reported `Environment.ProcessPath`?

You can overwrite the base directory with your own value as a last resort with `AppContext.SetData` but I'm surprised this would be necessary:

https://github.com/dotnet/runtime/blob/7faabde7066e9769de8dfa33164348efb791802f/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L26-L30

**Comment by ladeak** [comment]
`Environment.ProcessPath` points to `C:\Users\ladea\AppData\Local\Programs\Microsoft VS Code\Code.exe`.

I will check if I can override this in VS Code or use the AppContext as suggested. Thank you for the details.

**Comment by ladeak** [comment]
Are there any plans to solve this issue?
If someone has a single file publish (not even NativeAOT), it will fail to do HTTP/3 on both Linux and Windows.
For Linux users, there could be some instructions: setup MS package repository and the apt-get install msquic etc.. But on Windows, I don't know any such equivalent. I am not aware any official msquic builds, yet alone finding one for each platform. I would also need to know every time .NET runtime updates the msquic dependency, so I know when to bundle a new msquic file next to the executable.

If I build a single file executable on Windows and I would like to distribute it, I could "drop" the msquic file next to the executable to make it work.
But as soon as I want to support x86 and ARM64, I would need to bundle the platform specific msquic version making the packaging distribution more complex (as it kills the benefits of building a single file on the first place). 

Do you have any suggestions to make this work in a slightly nicer way?

Also it is completely un-intuitive. If I set `<Http3Support>true</Http3Support>`, why does it not work? It seems like an explicit setting to work around this problem, no?

**Comment by wfurt** [comment]
We should look into it for 12. Starting with upcoming 2.7, The linux version does not need to link copy of quictls so the extra size drops 10x. . (works only on distro with OpenSSL 3.5+)
